### PR TITLE
Add support for duplicate fieldnames in dom.Value#Struct

### PR DESCRIPTION
**Pull request by desaikd**
*Description of changes:*
This PR is to support duplicate fields in a Struct according to the [Ion Spec](https://amzn.github.io/ion-docs/docs/spec.html#struct).

*Tasks:*

1.  Added changes in the `dom.Value#Struct` for supporting duplicate fields. (All the `Struct` field will now be stored in `(fieldname, values)` -> `[string, Value[]]` form)
2.  `dom.Value#Struct` constructor would accept both `Iterable<string, Value>` as well as `Iterable<string, Value[]>` for creating `Struct`.
2.  Added `getAll` and `allFields` methods for getting all values of Struct fields.
3.  `toString` method returns with all the values of Struct fields.
2.  Changes in test suite for duplicate fields.
3.  add tests for duplicate fields.

### CODE REVIEW

File: src/dom/Null.ts [hunk 1]
```diff
@@ -120,7 +120,7 @@ export class Null extends Value(Object, IonTypes.NULL, FromJsConstructor.NONE) {
     this._unsupportedOperationOrNullDereference("fieldNames");
   }
 
-  fields(): [string, Value][] {
+  fields(): [string, Value[]][] {
```
Comment (zslayton) [inline src/dom/Null.ts:123]: Unfortunately, we can't change this method signature, the `Struct` constructor's signature, or how iteration works as those would be breaking changes requiring a new major version.

I think we should leave those as-is and add new functions that allow users to opt-in to the array format. For example:

```ts
.get(name)
// vs
.getAll(name)

.fields()
// vs
.allFields()

for (let [fieldName, fieldValue] of s) {/*...*/}
// vs
for (let [fieldName, fieldValueList] of s.allFields()) {/*...*/}
```

While I think `get()`/`getAll()` is a clear single-value/list-of-values name pair, I'm less enthusiastic about `fields()`/`allFields()`. Open to suggestions there. `fields()`/`fieldsWithAllValues` is more accurate but pretty verbose.

Iteration is interesting, because it's _possible_ to support multiple fields given the existing signatures. For example, the constructor takes an iterator:

```ts
constructor(fields: Iterable<[string, Value]>, annotations: string[] = []) { /*...*/ }
```

The `fields` iterable could yield `["foo", 4]` and then `["foo", 6]`. I think the constructor can/should support that, with a later call to `.getAll("foo")` returning `[4, 6]`.

On the other hand, I _don't_ think `Struct`'s iterator should produce the same behavior because that would be a breaking change. Some number of folks may already be relying on deduplication in their iteration. I think `Struct`'s iterator should stay the same; users interested in all values can iterate over `myStruct.allFields()` (or whatever we name it) instead.

File: test/dom/equivalence.ts [hunk 2]
```diff
@@ -347,4 +347,13 @@ describe("Equivalence", () => {
       )
     );
   });
+  it("equals() for Struct with duplicate fields", () => {
+    let value1: Value = load("{ foo: 'bar', foo: 'baz' }")!;
+    let value2: Value = load("{ foo: 'bar', foo: 'baz' }")!;
+    let value3: Value = load("{ foo: 'bar', foo: 'qux' }")!;
+
+    assert.isTrue(value1.equals(value2));
+    assert.isTrue(value1.equals(value2));
```
Comment (zslayton) [inline test/dom/equivalence.ts:356]: This line is a dup of the one above it.

File: test/dom/equivalence.ts [hunk 3]
```diff
@@ -347,4 +347,13 @@ describe("Equivalence", () => {
       )
     );
   });
+  it("equals() for Struct with duplicate fields", () => {
+    let value1: Value = load("{ foo: 'bar', foo: 'baz' }")!;
+    let value2: Value = load("{ foo: 'bar', foo: 'baz' }")!;
+    let value3: Value = load("{ foo: 'bar', foo: 'qux' }")!;
```
Comment (zslayton) [inline test/dom/equivalence.ts:353]: Can you add a test that verifies that field order is not considered for equality? e.g.

```ts
{foo: 1, baz: true, foo: 2}
// is equal to
{foo: 2, foo: 1, baz: true}
```

File: test/dom/dom.ts [hunk 4]
```diff
@@ -484,12 +484,37 @@ describe('DOM', () => {
...
+          'age: 55, ' +
+          'age: 41' +
+          '}'
+      )!;
+
+      assert.equal(IonTypes.STRUCT, s.getType());
+      assert.deepEqual(['foo', 'bar'], s.getAnnotations());
+
+      // Field access return last value for a fieldname
+      assert.equal(41, s.age);
```
Comment (zslayton) [inline test/dom/dom.ts:508]: We'll need a test of the `getAll` behavior when it's available.

File: src/dom/Null.ts (hunk 1 above)
Comment (zslayton) [reply src/dom/Null.ts:123]: I suppose the constructor could also accept a `Iterable<[string, Value]> | Iterable<[string, Value[]]>` to be more flexible.

**Comment by desaikd** [comment]
*Completed Tasks:*

1. added changes for `getAll()` and `allFields()` methods.
2. changed `ionEquals()` method for checking unordered duplicate fields.
2. Tests suite changes for all the above tasks.

*Questions:*
Could you explain with an example on how the `fields` return value would look like? Also for the iterator?
e.g.
```
{foo: 1, baz: true, foo: 2} 
```
For above Struct value what should be the return value for `fields()` as well as iterator ? 
Currently it will return something like this:
```
[['foo',[1,2]] , ['baz' , true]]
```

**Comment by desaikd** [comment]
*Changes:*
```
{foo: 1, baz: true, foo: 2} 
```
`fields()` would return Struct fields as below with only the last value:
```
[['foo', 2] , ['baz' , true]]
```

- To get all the field values `allFields()` can be used.
- `toString` method returns with all the values of Struct fields.

File: src/dom/Struct.ts [hunk 5]
```diff
@@ -46,8 +46,11 @@ export class Struct extends Value(
     annotations: string[] = []
   ) {
     super();
+    // Struct fields will always be stored as (fieldName, fieldValues) => (string, Value[])
     for (const [fieldName, fieldValue] of fields) {
-      this._fields[fieldName] = fieldValue;
+      this._fields[fieldName] = Array.isArray(fieldValue)
```
Comment (zslayton) [inline src/dom/Struct.ts:51]: `dom.List` and `dom.SExpression` inherit from `Array` -- is this able to distinguish between an `Array` and a `Value` that happens to inherit from `Array`?

Comment (desaikd) [reply src/dom/Struct.ts:51]: No. Its doesn't distinguish. Maybe I can reverse the check and see if it's an `instanceof Value` first and otherwise it will always be `Value[]` array because of the constructor arguments.

File: src/dom/Value.ts [hunk 6]
```diff
@@ -106,6 +106,14 @@ export interface Value {
    */
   get(...pathElements: PathElement[]): Value | null;
 
+  /**
+   * For the Struct type, returns an array containing all the values for given field name or field name path
```
Comment (zslayton) [inline src/dom/Value.ts:110]: We should emphasize that this is meant to be used when working with input that contains duplicate fields. Maybe:

```suggestion
   * Like the `get(PathElement[])` method, but returns *all* of the values associated 
   * with a given field name instead of just the last one. For example, for the
   * following struct:
   * 
   *     {
   *         foo: 1,
   *         foo: null,
   *         foo: true,
   *     }
   * 
   * a call to `get("foo")` would return a `dom.Boolean` representing `true`
   * while a call to `getAll("foo")` would return an `Array` of
   * `dom.Value`s: `[1, null, true]`.
```

File: src/dom/Value.ts [hunk 7]
```diff
@@ -118,6 +126,12 @@ export interface Value {
    */
   fields(): [string, Value][];
 
+  /**
+   * For the Struct type, returns an array containing the field name/values pairs in the Struct;
```
Comment (zslayton) [inline src/dom/Value.ts:130]: Like `get()`/`getAll()`, I would describe this function in its relation to `fields()`.

**Review by zslayton — APPROVED** [review]
